### ISSUE TITLE
Optimization for first-screen rendering time

### ISSUE THREAD

**Issue by Jinjiang** [weexteam/vue-framework-for-Apache-Weex#2]
Transferred from https://github.com/alibaba/weex/issues/363

test case:
- Vue: https://github.com/alibaba/weex/blob/88422ee8567bb946d87d465446137916ac9dd605/src/vue/examples/showcase/itemlist.vue (using weex-vue-loader to build JS bundle)
- Weex: https://github.com/alibaba/weex/blob/88422ee8567bb946d87d465446137916ac9dd605/examples/component/list/list-demo.we (using weex-loader to build JS bundle)

The `screenRenderTime` is really important for ux. So it's better to send DOM commands to native renderer as early as the framework generate elements.

**Comment by Jinjiang** [comment]
@cxfeng1 could you please leave some native benchmark tips here :-)

**Comment by cxfeng1-zz** [comment]
- Please follow the [iOS Getting Started Guide](https://github.com/alibaba/weex#ios) to start weex demo project in Xcode.
- While entering any weex page, you can see all the performance data in Xcode's console,  `screenRenderTime` and `totalTime` is more important than others.
- list-demo's screenRenderTime  is about 100ms on iPhone6,  250ms on iPhone5.

**Comment by yyx990803** [comment]
The root cause of the slower initial render time is largely due to how Vue's vdom patch algorithm works: by default, the patching algorithm appends the whole tree from bottom up - that is, leaves are appended into their parents first (in memory), and actual render only happens when the root node is finally appended. So when there are long lists being rendered, the user won't see anything until the whole vdom tree has been patched and appended.

To illustrate, say we have a list of 10 items, each contains a single `<text>` node. This is what happens:

```
    - append text to item 0 (in memory)
  - append item 0 to list (in memory)
  - ...
    - append text to item 9 (in memory)
  - append item 9 to list (in memory)
- append list to document (native render)
```

In the [stream-render branch](https://github.com/weexteam/weex-vue-framework/commits/stream-render) I have implemented render perf improvements by adjusting vnode patching algorithm so that:
- When a vnode has more than a given threshold of children (e.g. 10 - this may be subject to change based on more profiling in real apps), it is considered a "long list".
- For each "long list" node, we will reverse the default Node appending order so that we **first append the parent node, and then append the list items**. This gives us the following render order:

```
- append list to document (native render)
    - append text to item 0 (in memory)
  - append item 0 to list (native render)
  - ...
    - append text to item 9 (in memory)
  - append item 9 to list (native render)
```

As you can see, assuming the whole process takes the same amount of time in JavaScript, with the adjusted algorithm the native renderer gets to render-as-we-go without having to wait for the whole process to finish. This results in faster initial render.

The reason for not reversing append order for all nodes is because each native render call has a communication cost. By only reversing append order for long lists, we ensure that we issue only 1 render call for each item in the list (notice the "append text to item" operation happens in memory). Imagine if each item contains more complex nested content, we will end up with way too many render calls if we reverse order for all nodes.

I have recored the perf numbers in the following google spreadsheet (data collected by running the list demo in iOS simulator):

https://docs.google.com/spreadsheets/d/1ZNNYY3wwsCb11xx8yaOCLW7f1Z1iiZ-kZa-zM24YNm4/edit?usp=sharing

**Comment by yyx990803** [comment]
implemented in weex-vue-framework@2.1.3-weex.1
